### PR TITLE
adding Link rate statistics out side UI

### PR DESCRIPTION
**Pull request by chandrasheker**
Fix for Link rate https://github.com/Juniper/warp17/issues/12

### CODE REVIEW

File: src/tpg_test_mgmt_cli.c [hunk 1]
```diff
@@ -105,19 +105,19 @@ struct cmd_tests_start_stop_result {
 };
 
 static cmdline_parse_token_string_t cmd_tests_start_T_start =
-    TOKEN_STRING_INITIALIZER(struct cmd_tests_start_stop_result, start, "start");
+TOKEN_STRING_INITIALIZER(struct cmd_tests_start_stop_result, start, "start");
```
Comment (dceara) [inline src/tpg_test_mgmt_cli.c]: Please follow the initial alignment and indentation style of the file. This applies to the whole pull request.

The main guidelines are:
- use spaces not tabs.
- the continuation of a statement on a new line should be indented with 4 spaces.
- function arguments both for function definitions and function calls should be aligned with the open parenthesis `(`.
- align variable definitions that are in the same block so the variable names (including `*`) are aligned:

    ```
    int    a;
    float *p;
    long   b;
    
    long unsigned x;
    double        y;
    ```

File: src/tpg_test_mgmt_cli.c [hunk 2]
```diff
@@ -186,6 +186,78 @@ cmdline_parse_inst_t cmd_tests_stop = {
...
+/****************************************************************************
+ * - "show link rate"
+ ****************************************************************************/
+struct cmd_show_link_rate_result {
+    cmdline_fixed_string_t show;
+    cmdline_fixed_string_t link;
+    cmdline_fixed_string_t rate;
+};
+
+static cmdline_parse_token_string_t cmd_show_link_rate_T_show =
```
Comment (dceara) [inline src/tpg_test_mgmt_cli.c]: Please indent statement continuations with 4 spaces.

File: src/tpg_test_mgmt_cli.c [hunk 3]
```diff
@@ -186,6 +186,78 @@ cmdline_parse_inst_t cmd_tests_stop = {
...
+TOKEN_STRING_INITIALIZER(struct cmd_show_link_rate_result, show, "show");
+static cmdline_parse_token_string_t cmd_show_link_rate_T_link =
+TOKEN_STRING_INITIALIZER(struct cmd_show_link_rate_result, link, "link");
+static cmdline_parse_token_string_t cmd_show_link_rate_T_rate =
+TOKEN_STRING_INITIALIZER(struct cmd_show_link_rate_result, rate, "rate");
+
+static void tpg_print_line_rate(void)
+{
+    int i = 0;
```
Comment (dceara) [inline src/tpg_test_mgmt_cli.c]: Please align variable names for variables defined in the same block.

File: src/tpg_test_mgmt_cli.c [hunk 4]
```diff
@@ -186,6 +186,78 @@ cmdline_parse_inst_t cmd_tests_stop = {
...
+    int port = 0;
+    struct rte_eth_link link_info;
+    struct rte_eth_stats link_rate_stats;
+    double    usage_tx;
+    double    usage_rx;
+    uint64_t  link_speed_bytes;
+
+    port = rte_eth_dev_count() + ring_if_get_count();
+
+    for(i=0;i<port;i++){
```
Comment (dceara) [inline src/tpg_test_mgmt_cli.c]: Missing spaces. Should be `for (i = 0; i < port; i++) {`

File: src/tpg_test_mgmt_cli.c [hunk 5]
```diff
@@ -186,6 +186,78 @@ cmdline_parse_inst_t cmd_tests_stop = {
...
+        link_speed_bytes = (uint64_t)link_info.link_speed * 1000 * 1000 / 8;
+
+        if (link_info.link_status) {
+            usage_tx = (double)link_rate_stats.obytes * 100 / link_speed_bytes;
+            usage_rx = (double)link_rate_stats.ibytes * 100 / link_speed_bytes;
+        } else {
+            usage_tx = 0;
+            usage_rx = 0;
+        }
+        printf( "Port %"PRIu32": link %s, speed %d%s, "
```
Comment (dceara) [inline src/tpg_test_mgmt_cli.c]: Extra whitespace after `(`.

File: src/tpg_test_mgmt_cli.c [hunk 6]
```diff
@@ -186,6 +186,78 @@ cmdline_parse_inst_t cmd_tests_stop = {
...
+};
+
+static cmdline_parse_token_string_t cmd_show_link_rate_T_show =
+TOKEN_STRING_INITIALIZER(struct cmd_show_link_rate_result, show, "show");
+static cmdline_parse_token_string_t cmd_show_link_rate_T_link =
+TOKEN_STRING_INITIALIZER(struct cmd_show_link_rate_result, link, "link");
+static cmdline_parse_token_string_t cmd_show_link_rate_T_rate =
+TOKEN_STRING_INITIALIZER(struct cmd_show_link_rate_result, rate, "rate");
+
+static void tpg_print_line_rate(void)
```
Comment (dceara) [inline src/tpg_test_mgmt_cli.c]: This code will now be duplicated in tpg_test_mgmt_cli.c and tpg_test_stats.c. Isn't it better to add a function that displays the link stats on a generic printer. For example, the per test statistics are dumped with:

```
void test_state_show_stats(const tpg_test_case_t *te,
                           printer_arg_t *printer_arg);
```

Why not define a function for printing link rate in a generic way and call that one both from the CLI (in `cmd_show_link_rate_parsed`) and from the UI (in `test_display_stats_hdr`)? A prototype could be:

```
void test_show_link_rate(uint32_t port, printer_arg_t *printer_arg);
```

Comment (chandrasheker) [inline src/tpg_test_mgmt_cli.c]: With proposed prototype of test_show_link_rate, How can we have two different kinds of printing formats in test_show_link_rate for CLI(tpg_printf) and UI(UI_PRINTLN_WIN) which accepts different arguments to have generic way of printing?

Comment (dceara) [inline src/tpg_test_mgmt_cli.c]: I moved test_show_link_rate in tpg_test_stats.c and called it with a`printer_arg_t  parg = TPG_PRINTER_ARG(ui_printer, win);` argument. That should do the trick.

**Pull request merged**